### PR TITLE
Implement Hilbert Transform Indicator

### PR DESCRIPTION
**Pull request by alexander-myltsev**
#### Description
Implements Hilbert Transform Indicator.

#### Related Issue
Closes https://github.com/QuantConnect/Lean/issues/6986

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

### CODE REVIEW

File: Indicators/HilbertTransformIndicator.cs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    private const int Quad2Length = 2;
+    private const int InPhase3Length = 3;
+
+    private readonly IndicatorBase<IndicatorDataPoint> _input;
+    private readonly IndicatorBase<IndicatorDataPoint> _prev;
+    private readonly IndicatorBase<IndicatorDataPoint> _v1;
+    private readonly IndicatorBase<IndicatorDataPoint> _v2;
+    private readonly IndicatorBase<IndicatorDataPoint> _v4;
+    private readonly Queue<IndicatorDataPoint> _inPhase3 = new();
+    private readonly Queue<IndicatorDataPoint> _quad2 = new();
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:25]: I suggest replacing the `Queue` for `RollingWindow`.
```csharp
private readonly RollingWindow<IndicatorDataPoint> _quadratureWindow = new RollingWindow<IndicatorDataPoint>(2);
```
where `2` comes from `Quad2Length = 2` and we can remove this field. Same for _inPhase3.

File: Indicators/HilbertTransformIndicator.cs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    /// <summary>
+    /// Real (inPhase) part of complex number component of price values
+    /// </summary>
+    public IndicatorBase<IndicatorDataPoint> InPhase { get; }
+
+    /// <summary>
+    /// Imaginary (quadrature) part of complex number component of price values
+    /// </summary>
+    public IndicatorBase<IndicatorDataPoint> Quad { get; }
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:39]: Please avoid abbreviations. We can name this property `Quadrature`.

File: Indicators/HilbertTransformIndicator.cs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    /// <param name="name">The name of this indicator</param>
+    /// <param name="length">The length of the FIR filter used in the calculation of the Hilbert Transform.
+    /// This parameter determines the number of filter coefficients in the FIR filter.</param>
+    /// <param name="iMult">The multiplication factor used in the calculation of the in-phase component
+    /// of the Hilbert Transform. This parameter adjusts the sensitivity and responsiveness of
+    /// the transform to changes in the input signal.</param>
+    /// <param name="qMult">The multiplication factor used in the calculation of the quadrature component of
+    /// the Hilbert Transform. This parameter also adjusts the sensitivity and responsiveness of the
+    /// transform to changes in the input signal.</param>
+    public HilbertTransformIndicator(string name, int length, decimal iMult, decimal qMult)
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:53]: Nit/style. In Lean, we avoid Hungarian notation or abbreviation, so `iMult` should be `inPhaseMultiplicationFactor` or `inPhaseMultiplier`. Same for `qMult`.

File: Indicators/HilbertTransformIndicator.cs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        : base(name)
+    {
+        _quad2WarmUpPeriod = length + 2;
+        _inPhase3WarmUpPeriod = _quad2WarmUpPeriod + InPhase3Length;
+
+        _input = new Identity(name + "_input");
+        _prev = new Delay(name + "_prev", length);
+        _v1 = new FunctionalIndicator<IndicatorDataPoint>(name + "_v1",
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:61]: `_v1` looks like a simple difference, so we could use the `Minus` extension:
```csharp
_change = _prev.Minus(_input);
```

Perhaps `_v1` , `_v2` , and `_v4` could have descriptive names, so we know what they are trying to calculate.

File: Tests/Indicators/HilbertTransformIndicatorTests.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+[TestFixture]
+public class HilbertTransformIndicatorTests : CommonIndicatorTests<IndicatorDataPoint>
+{
+    protected override IndicatorBase<IndicatorDataPoint> CreateIndicator()
+    {
+        return new HilbertTransformIndicator();
+    }
+
+    protected override string TestFileName => "spy_with_hilbert.csv";
```
Comment (AlexCatarino) [inline Tests/Indicators/HilbertTransformIndicatorTests.cs:30]: What is the source of this file? Did you generate it or can we find its content at  http://www.technicalanalysis.org.uk/moving-averages/Ehle.pdf?

File: Indicators/HilbertTransformIndicator.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
+using System;
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:1]: Please add QuantConnect's Apache License headers to cs files.

File: Indicators/HilbertTransformIndicator.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    /// <summary>
+    /// Resets this indicator to its initial state
+    /// </summary>
+    public override void Reset()
+    {
+        base.Reset();
+        InPhase.Reset();
+        Quad.Reset();
+        updatesCount = 0;
```
Comment (AlexCatarino) [inline Indicators/HilbertTransformIndicator.cs:160]: All the indicators `_prev`, `input`, `_v1` , `_v2` ,  and `_v4`  should call Reset.

**Review by AlexCatarino — CHANGES_REQUESTED** [review]
Thank you for the contribution, @alexander-myltsev !
I left some comments on naming conventions and leveraging the Indicators API.

I would like to request additional comments on the internal indicators. You could the formulas from the articles and briefly explain what they mean.

This pull request doesn't include the helper method implementation in QCAlgorithm.

File: Indicators/HilbertTransformIndicator.cs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+    private readonly IndicatorBase<IndicatorDataPoint> _input;
+    private readonly IndicatorBase<IndicatorDataPoint> _prev;
+    private readonly IndicatorBase<IndicatorDataPoint> _v1;
+    private readonly IndicatorBase<IndicatorDataPoint> _v2;
+    private readonly IndicatorBase<IndicatorDataPoint> _v4;
+    private readonly Queue<IndicatorDataPoint> _inPhase3 = new();
+    private readonly Queue<IndicatorDataPoint> _quad2 = new();
+
+    private int updatesCount = 0;
```
[the remark this answers is not in the document]
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs:27]: fixed by 7344992cc14234a6716d53c9727ff982b7818cf8

[the remark this answers is not in the document]
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixed by 7344992cc14234a6716d53c9727ff982b7818cf8

File: Indicators/HilbertTransformIndicator.cs (hunk 2 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixed by 67cbc51ec4f762a4ac89b19c12a9576408995d98

File: Indicators/HilbertTransformIndicator.cs (hunk 3 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixed by 67cbc51ec4f762a4ac89b19c12a9576408995d98

File: Indicators/HilbertTransformIndicator.cs (hunk 4 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixed by 7344992cc14234a6716d53c9727ff982b7818cf8

File: Tests/Indicators/HilbertTransformIndicatorTests.cs (hunk 5 above)
Comment (alexander-myltsev) [reply Tests/Indicators/HilbertTransformIndicatorTests.cs:30]: I did see that J.Ehlers provides data that can be used for testing. I generated it myself from the code https://github.com/ooples/OoplesFinance.StockIndicators/blob/be83f40b547301aae7d0196b1255f4eed38d61e9/src/Calculations/Ehlers.cs#L3781-L3823

File: Indicators/HilbertTransformIndicator.cs (hunk 6 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs:16]: fixed by ffd25a4d0dcac7f84e52f2b616fd42ad4a437119

File: Indicators/HilbertTransformIndicator.cs (hunk 1 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixed by bb5b228f9f3c1fe6d2df4b487cdbcd4c08a501d7

File: Indicators/HilbertTransformIndicator.cs (hunk 7 above)
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs]: fixes b005a02c353facb021e864c5780267125222dbc1

File: Tests/Indicators/HilbertTransformIndicatorTests.cs (hunk 5 above)
Comment (Martin-Molinero) [reply Tests/Indicators/HilbertTransformIndicatorTests.cs:30]: I believe LS shared values for each indicator to assert against at https://github.com/QuantConnect/Lean/issues/6986#issuecomment-1441554170 should use these 👌

**Review by Martin-Molinero — COMMENTED** [review]
Looks nice! 
Please take a look at the following requested changes [review.patch](https://github.com/QuantConnect/Lean/files/11140681/review.patch). Unit tests should assert against LS shared expected values https://github.com/QuantConnect/Lean/issues/6986#issuecomment-1441554170

File: Indicators/HilbertTransformIndicator.cs [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+    /// the transform to changes in the input signal.</param>
+    /// <param name="quadratureMultiplicationFactor">The multiplication factor used in the calculation of the quadrature component of
+    /// the Hilbert Transform. This parameter also adjusts the sensitivity and responsiveness of the
+    /// transform to changes in the input signal.</param>
+    public HilbertTransformIndicator(string name, int length, decimal inPhaseMultiplicationFactor, decimal quadratureMultiplicationFactor)
+        : base(name)
+    {
+        var quadratureWarmUpPeriod = length;
+        var inPhaseWarmUpPeriod = length + 2;
+        WarmUpPeriod = Math.Max(quadratureWarmUpPeriod, inPhaseWarmUpPeriod);
```
[the remark this answers is not in the document]
Comment (alexander-myltsev) [reply Indicators/HilbertTransformIndicator.cs:65]: There will be values for length 1:

```
date,close,prev,detrendPrice,detrendPriceDelay2,detrendPriceDelay4,inPhaseDelay3,inPhase,quadratureDelay2,quadrature
1/3/2022 12:00:00 AM,4796.56,0,0,0,0,0,0,0,0
1/4/2022 12:00:00 AM,4793.54,4796.56,-3.0200000000004366,0,0,0,0,0,1.0207600000001475
1/5/2022 12:00:00 AM,4700.58,4793.54,-92.96000000000004,0,0,0,0,0,31.420480000000016
1/6/2022 12:00:00 AM,4696.05,4700.58,-4.529999999999745,-3.02,0,0,2.397125,1.0208,-1.143829600000086
1/7/2022 12:00:00 AM,4677.03,4696.05,-19.020000000000437,-92.96,0,0,73.78699999999999,31.4205,-75.91111099999983
1/10/2022 12:00:00 AM,4670.29,4677.03,-6.739999999999782,-4.53,-3.02,0,-0.17931249999999954,-1.1438,-2.638484400000074
1/11/2022 12:00:00 AM,4713.07,4670.29,42.779999999999745,-19.02,-92.96,2.3971,-99.58071649999998,-75.9111,-59.137591799999925
1/12/2022 12:00:00 AM,4726.35,4713.07,13.280000000000655,-6.74,-4.53,73.787,46.542120000000004,-2.6385,-12.120453000000223
1/13/2022 12:00:00 AM,4659.03,4726.35,-67.32000000000062,42.78,-19.02,-0.1793,-57.845480499999994,-59.1376,45.54565120000021
1/14/2022 12:00:00 AM,4662.85,4659.03,3.8200000000006185,13.28,-6.74,-99.5807,-82.1997445,-12.1205,7.89211099999979
1/18/2022 12:00:00 AM,4577.11,4662.85,-85.74000000000069,-67.32,42.78,46.5421,136.4644835,45.5457,-22.945433399999757
```

But they are only reasonable after `inPhaseDelay3` or `_quadratureDelay2` is ready.

I don't see the problem with `length=1`. `_detrendPrice` would be ready on step 2. `_detrendPriceDelay4` would be on step 6. `quadrature` can be computed. Where's the problem here?

Also note, that `inPhase` and `quadrature` are ready in different times.

**Comment by LouisSzeto** [comment]
@alexander-myltsev 

The SPY data used: 
[spy.csv](https://github.com/QuantConnect/Lean/files/11199121/spy.csv)

Base code:
```
qb = QuantBook()

spy = qb.AddEquity("SPY", Resolution.Daily).Symbol
qqq = qb.AddEquity("QQQ", Resolution.Daily).Symbol

history = qb.History([spy, qqq], datetime(1998, 1, 1), datetime(2023, 2, 23))
ret = history.close.pct_change().unstack(0).dropna()

close = history.loc[spy].close
high = history.loc[spy].high
low = history.loc[spy].low
volume = history.loc[spy].volume
```

**Comment by alexander-myltsev** [comment]
JB rejected my contribution. I won't finish.

**Pull request closed without merging**